File: firmware/src/store.cpp
#include "store.h"

#include "LittleFS.h"

Store::Store(int motorId, Settings* settings) {
  _motorId = motorId;
  _settings = settings;
}

void Store::begin() {
  File f = LittleFS.open("settings_" + String(_motorId), "r");
  uint8_t buffer[15];
  if (f) {
    f.read(buffer, 15);
    f.close();
    _settings->parse_settings_from_bytes(buffer, 15);
  }  
}

void Store::storeSettingsFromBytes(uint8_t* bytes) {
  _settings->parse_settings_from_bytes(bytes, 15);
  persistSettings();
}

int Store::writeSettingsToBytes(uint8_t* buffer) {
  _settings->get_settings_as_bytes(buffer);
  return 15;
}

int32_t Store::getLastPosition() {
  File f = LittleFS.open("position_" + String(_motorId), "r");
  int32_t pos;
  if (f) {
    uint8_t buf[4];
    f.read(buf, 4);
    memcpy(&pos, &buf[0], 4);
  }
  cachedPosition = pos;
  return pos;
}

void Store::storeLastPosition(int32_t pos) {
  if (pos == cachedPosition) return; 
  File f = LittleFS.open("position_" + String(_motorId), "w");
  uint8_t buf[4];
  memcpy(&buf[0], &pos, 4);
  f.write(buf, 4);
  f.close();
}

void Store::persistSettings() {
  uint8_t buffer[15];
  _settings->get_settings_as_bytes(buffer);
  File f = LittleFS.open("settings_" + String(_motorId), "w");
  f.write(buffer, 15);
  f.close();
}

void Store::storeBottomLimit(int32_t bottomLimit) {
  Serial.printf("Storing bottom limit: %d\n", bottomLimit);
  _settings->bottom_limit = bottomLimit;
  persistSettings();
}

int32_t Store::getBottomLimit() {
  return _settings->bottom_limit;
}

void Store::storeDefaultMaxSpeed(float defaultMaxSpeed) {
  _settings->default_max_speed = defaultMaxSpeed;
  persistSettings();
}
float Store::getDefaultMaxSpeed() {
  return _settings->default_max_speed;
}

int Store::getSlowdownPercent() {
  return _settings->slowdown_percent;
}
int Store::getManualSpeedUp() {
  return _settings->manual_speed_up;
}
int Store::getManualSpeedDown() {
  return _settings->manual_speed_down;
}
int Store::getPidKp() {
  return _settings->pid_kp;
}
int Store::getPidKd() {
  return _settings->pid_kd;
}
int Store::getPidKi() {
  return _settings->pid_ki;
}

File: firmware-micropython/app/store.py
import app.settings

def read_last_position(motor):
    try:
        with open(f"position_{motor}", "r") as f:
            return int(f.read())
    except OSError:
        return 0

def store_position(motor, position):
    with open(f"position_{motor}", "w") as f:
        f.write(f"{position}")

def read_settings(motor):
    try:
        with open(f"settings_{motor}", "rb") as f:
            return f.read()
    except OSError:
        return []

def store_settings(motor, settings):
    with open(f"settings_{motor}", "wb") as f:
        f.write(settings)

class Store:

  def __init__(self, motor):
    self._motor = motor
    self._last_position = read_last_position(motor)
    self._settings = app.settings.Settings(read_settings(motor))

  def store_last_position(self, last_position):
    if last_position != self._last_position: 
      print(f'storing last position {self._motor}: {last_position}')
      self._last_position = last_position
      store_position(self._motor, self._last_position)

  def get_last_position(self):
    return self._last_position

  def store_bottom_limit(self, bottom_limit):
    print(f'storing bottom position {bottom_limit} on motor {self._motor}')
    self._settings.bottom_limit = bottom_limit
    store_settings(self._motor, self._settings.get_settings_as_bytes())

  def get_slowdown_percent(self):
    return self._settings.slowdown_percent

  def get_bottom_limit(self):
    return self._settings.bottom_limit

  def get_manual_speed_up(self):
    return self._settings.manual_speed_up

  def get_manual_speed_down(self):
    return self._settings.manual_speed_down

  def store_default_max_speed(self, default_max_speed):
    print(f'storing default max speed {default_max_speed} on motor {self._motor}')
    self._settings.default_max_speed = default_max_speed
    store_settings(self._motor, self._settings.get_settings_as_bytes())

  def get_default_max_speed(self):
    return self._settings.default_max_speed

  def get_pid_kp(self):
    return self._settings.pid_kp

  def get_pid_kd(self):
    return self._settings.pid_kd

  def get_pid_ki(self):
    return self._settings.pid_ki

  def store_settings_from_string(self, as_string):
    self._settings.parse_settings_from_string(as_string)
    store_settings(self._motor, self._settings.get_settings_as_bytes())

  def get_settings_as_string(self):
    return self._settings.get_settings_as_string()  